gh-69753: Add Bytes Object Support to Shlex

HassanAbouelela opened the pull request:
Adds support for bytes objects in the shlex module.

<!-- issue-number: [bpo-25567](https://bugs.python.org/issue25567) -->
https://bugs.python.org/issue25567
<!-- /issue-number -->


<!-- gh-issue-number: gh-69753 -->
* Issue: gh-69753
<!-- /gh-issue-number -->

HassanAbouelela committed:
Add Bytes Support to shlex

Adds checking and conversions for bytes objects, to allow them to be used with shlex.

HassanAbouelela committed:
Add Tests for shlex

Adds tests that cover bytestrings in the shlex module.

HassanAbouelela committed:
Add News Blurb

HassanAbouelela committed:
Return Bytes from Split

Allows split to return a list of bytes, given a byte input string.

JelleZijlstra commented on Lib/test/test_shlex.py:
```diff
@@ -339,6 +343,22 @@ def testQuote(self):
+        safeunquoted = string.ascii_letters + string.digits + '@%_-+=:,./'
+        unsafe = '"`$\\!'
+
+        self.assertEqual(shlex.quote(b''), b"''")
```
Would be good to test some bytestrings with non-ASCII characters, to make sure that they're handled correctly.

HassanAbouelela replied:
I added the missing unicode characters from the `testQuote` test, and changes the encoding to utf-8 to account for the changes. Added in bfb76be9.

picnixz replied:
Please add *all* characters that are supported on POSIX (not just three of them):

```py
unicode_chars = ('ßàáâãäåæçèéêëìíîïðñòóôõöøùúûüýþÿ'
 				 'ÀÁÂÃÄÅÆÇÈÉÊËÌÍÎÏÐÑÒÓÔÕÖØÙÚÛÜÝÞ')
```

(I'll let you CC the chars since I am not sure about my own CC)

HassanAbouelela committed:
Merge branch 'master' into bpo-25567-shlex-bytestrings

HassanAbouelela committed:
Adds Non-Ascii Characters To Shlex Bytes Tests

Adds the unicode sample characters from the `testQuote` test to the
bytes version.

HassanAbouelela committed:
Merge branch 'main' into bpo-25567-shlex-bytestrings

MaxwellDupre reviewed (approved):
All shlex test run ok.
Also example in bugs look ok.

HassanAbouelela committed:
Merge branch 'main' into bpo-25567-shlex-bytestrings

serhiy-storchaka reviewed (commented):
I see several problems here:

* If `bytes` is passed to `shlex()`, should not iteration produce `bytes` rather of `str`?
* If `shlex()` accepts `str`, `bytes` and text files, should not it support also binary files?
* `join()` returns `str` for a sequence of `str` and `bytes` for a sequence of `bytes`. Unless the sequence is empty, in which case it always return `str`. The problem is that an empty sequence of `str` is the same as the empty sequence of `bytes`.

picnixz commented on Lib/shlex.py:
```diff
@@ -22,6 +22,10 @@ def __init__(self, instream=None, infile=None, posix=False,
             instream = StringIO(instream)
+        elif isinstance(instream, bytes):
+            # convert byte instreams to string
+            instream = StringIO(instream.decode("ascii", "surrogateescape"))
```
Why would you only support ASCII strings? for POSIX platforms, you should support also additional characters (see wordchars being extended).

picnixz commented on Lib/shlex.py:
```diff
@@ -312,26 +316,78 @@ def split(s, comments=False, posix=True):
-    return list(lex)
+
+    if isinstance(s, bytes):
+        return [i.encode("ascii") for i in lex]
```
Again, this should not be restricted to ASCII characters only.

picnixz commented on Lib/shlex.py:
```diff
@@ -312,26 +316,78 @@ def split(s, comments=False, posix=True):
 def join(split_command):
     """Return a shell-escaped string from *split_command*."""
-    return ' '.join(quote(arg) for arg in split_command)
+    if len(split_command) == 0:
```
This behaviour should be documented, saying that a string will be returned instead of bytes for an empty sequence of bytes. Or maybe add an additional a separate function which only accepts bytes inputs (less code and no warnings and more efficient).

picnixz commented on Lib/shlex.py:
```diff
@@ -312,26 +316,78 @@ def split(s, comments=False, posix=True):
+    cleaned = []
+    warned = False
+
+    for command in split_command:
```
I think it is too complicated. The str-version assumes that the objects are all strings, namely that `split_command` is an iterable of strings.

picnixz commented on Lib/shlex.py:
```diff
@@ -312,26 +316,78 @@ def split(s, comments=False, posix=True):
+        # use single quotes, and put single quotes into double quotes
+        # the string $'b is then quoted as '$'"'"'b'
+        return b"'" + s.replace(b"'", b"'\"'\"'") + b"'"
+
```
```suggestion
```

picnixz commented on Lib/test/test_shlex.py:
```diff
@@ -171,6 +171,10 @@ def testSplitPosix(self):
 
+    def testSplitBytes(self):
+        """Test byte objects splitting"""
+        self.assertEqual(shlex.split(b"split words"), [b"split", b"words"])
```
The coverage is insufficient. Use the `self.splitTest` as for the string case but do it for bytes inputs instead. 

In addition, use more than just ASCII characters but also those that are supported by the POSIX platforms.

picnixz commented on Lib/test/test_shlex.py:
```diff
@@ -358,6 +382,46 @@ def testJoinRoundtrip(self):
                 resplit = shlex.split(joined)
                 self.assertEqual(split_command, resplit)
 
+    def testJoinBytes(self):
```
Use the same dataset as for those used in strings, especially with quotation symbols or unsafe characters.

picnixz commented on Misc/NEWS.d/next/Library/2020-10-12-03-46-25.bpo-25567.xgfgij.rst:
```diff
@@ -0,0 +1 @@
+Add support for bytes objects in the shlex module.
```
```suggestion
Add support for :class:`bytes` objects to the :mod:`shlex` module.
```

You should probably explain that `join()` returns an empty string as well in the docs. This is important. Or introduce a separate function for bytes objects (which I think would be preferrable because you wouldn't have all those warnings to handle).
